Panic: Runtime error in SweepingProvider during concurrent trie access (addManyAtDepth/pruneSubtrieAtDepth)

### Checklist

- [x] I have searched the [issue tracker](https://github.com/libp2p/go-libp2p-kad-dht/issues?q=is%3Aissue) for existing issues.

### Description

Description
We are observing intermittent nil-pointer dereference panics in the SweepingProvider when running in a 3-node HA Kubernetes environment (K3s) with an embedded registry (Spegel) enabled. These panics occur in the underlying go-libdht trie logic during background maintenance cycles.

The crashes occur during periods of high system latency (e.g., etcd snapshots or database compaction), which appears to widen the race condition window in the recursive trie-walking logic.

Environment
- Library Version: v0.36.0 (as bundled in k3s v1.34.7)
- Dependency: github.com/probe-lab/go-libdht@v0.4.0
- Architecture: 3-node HA cluster 

Stack Traces:
Case A: Insertion Panic (batchReprovide)
Occurs during the background reprovide cycle while adding keys to the trie:
```
panic: runtime error: invalid memory address or nil pointer dereference
May 06 12:30:17 stks03 k3s[3402045]: panic: runtime error: invalid memory address or nil pointer dereference
May 06 12:30:17 stks03 k3s[3402045]: [signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x2fc03dd]
May 06 12:30:17 stks03 k3s[3402045]: goroutine 486059188 [running]:
May 06 12:30:17 stks03 k3s[3402045]: github.com/probe-lab/go-libdht/kad/trie.(*Trie[...]).IsLeaf(...)
May 06 12:30:17 stks03 k3s[3402045]:         /go/pkg/mod/github.com/probe-lab/go-libdht@v0.4.0/kad/trie/trie.go:59
May 06 12:30:17 stks03 k3s[3402045]: github.com/probe-lab/go-libdht/kad/trie.(*Trie[...]).addManyAtDepth(0x82cd860, 0x3, {0xc0185dc570, 0x1, 0x1})
May 06 12:30:17 stks03 k3s[3402045]:         /go/pkg/mod/github.com/probe-lab/go-libdht@v0.4.0/kad/trie/trie.go:159 +0x5d
May 06 12:30:17 stks03 k3s[3402045]: github.com/probe-lab/go-libdht/kad/trie.(*Trie[...]).addManyAtDepth(0x82cd860, 0x2, {0xc0185dc570, 0x1, 0x1})
May 06 12:30:17 stks03 k3s[3402045]:         /go/pkg/mod/github.com/probe-lab/go-libdht@v0.4.0/kad/trie/trie.go:220 +0x652
May 06 12:30:17 stks03 k3s[3402045]: github.com/probe-lab/go-libdht/kad/trie.(*Trie[...]).addManyAtDepth(0x82cd860, 0x1, {0xc0185dc570, 0x1, 0x1})
May 06 12:30:17 stks03 k3s[3402045]:         /go/pkg/mod/github.com/probe-lab/go-libdht@v0.4.0/kad/trie/trie.go:220 +0x652
May 06 12:30:17 stks03 k3s[3402045]: github.com/probe-lab/go-libdht/kad/trie.(*Trie[...]).addManyAtDepth(0x82cd860, 0x0, {0xc0185dc570, 0x1, 0x1})
May 06 12:30:17 stks03 k3s[3402045]:         /go/pkg/mod/github.com/probe-lab/go-libdht@v0.4.0/kad/trie/trie.go:220 +0x652
May 06 12:30:17 stks03 k3s[3402045]: github.com/probe-lab/go-libdht/kad/trie.(*Trie[...]).Add(...)
May 06 12:30:17 stks03 k3s[3402045]:         /go/pkg/mod/github.com/probe-lab/go-libdht@v0.4.0/kad/trie/trie.go:102
May 06 12:30:17 stks03 k3s[3402045]: github.com/libp2p/go-libp2p-kad-dht/provider/internal/timeseries.(*CycleStats).Add(0xc0020cc310, {0xc00000f900, 0x3}, 0x3af26c)
May 06 12:30:17 stks03 k3s[3402045]:         /go/pkg/mod/github.com/libp2p/go-libp2p-kad-dht@v0.36.0/provider/internal/timeseries/cycle_stats.go:62 +0x154
May 06 12:30:17 stks03 k3s[3402045]: github.com/libp2p/go-libp2p-kad-dht/provider.(*SweepingProvider).batchReprovide.func1()
May 06 12:30:17 stks03 k3s[3402045]:         /go/pkg/mod/github.com/libp2p/go-libp2p-kad-dht@v0.36.0/provider/provider.go:1734 +0x76
May 06 12:30:17 stks03 k3s[3402045]: github.com/libp2p/go-libp2p-kad-dht/provider.(*SweepingProvider).batchReprovide(0xc0020cc008, {0xc00000f900, 0x3})
May 06 12:30:17 stks03 k3s[3402045]:         /go/pkg/mod/github.com/libp2p/go-libp2p-kad-dht@v0.36.0/provider/provider.go:1748 +0x37e
May 06 12:30:17 stks03 k3s[3402045]: github.com/libp2p/go-libp2p-kad-dht/provider.(*SweepingProvider).reprovideLateRegions.func1({0xc00000f900?, 0x22?})
May 06 12:30:17 stks03 k3s[3402045]:         /go/pkg/mod/github.com/libp2p/go-libp2p-kad-dht@v0.36.0/provider/provider.go:1646 +0x2a
May 06 12:30:17 stks03 k3s[3402045]: created by github.com/libp2p/go-libp2p-kad-dht/provider.(*SweepingProvider).reprovideLateRegions in goroutine 4747179
May 06 12:30:17 stks03 k3s[3402045]:         /go/pkg/mod/github.com/libp2p/go-libp2p-kad-dht@v0.36.0/provider/provider.go:1645 +0x210
May 06 12:30:18 stks03 systemd[1]: k3s.service: Main process exited, code=exited, status=2/INVALIDARGUMENT
```

Case B: Deletion Panic (pruning)
```
Apr 18 15:58:33 stks03 k3s[1785244]: panic: runtime error: invalid memory address or nil pointer dereference
Apr 18 15:58:33 stks03 k3s[1785244]: [signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x2fd24be]
Apr 18 15:58:33 stks03 k3s[1785244]: goroutine 61351079 [running]:
Apr 18 15:58:33 stks03 k3s[1785244]: github.com/probe-lab/go-libdht/kad/trie.(*Trie[...]).HasKey(...)
Apr 18 15:58:33 stks03 k3s[1785244]:         /go/pkg/mod/github.com/probe-lab/go-libdht@v0.4.0/kad/trie/trie.go:54
Apr 18 15:58:33 stks03 k3s[1785244]: github.com/probe-lab/go-libdht/kad/trie.(*Trie[...]).IsEmptyLeaf(...)
Apr 18 15:58:33 stks03 k3s[1785244]:         /go/pkg/mod/github.com/probe-lab/go-libdht@v0.4.0/kad/trie/trie.go:64
Apr 18 15:58:33 stks03 k3s[1785244]: github.com/libp2p/go-libp2p-kad-dht/provider/internal/keyspace.pruneSubtrieAtDepth[...](0xc025c23bc0, {0xc002a81d00, 0x3}, 0x2)
Apr 18 15:58:33 stks03 k3s[1785244]:         /go/pkg/mod/github.com/libp2p/go-libp2p-kad-dht@v0.36.0/provider/internal/keyspace/trie.go:260 +0x13e
Apr 18 15:58:33 stks03 k3s[1785244]: github.com/libp2p/go-libp2p-kad-dht/provider/internal/keyspace.pruneSubtrieAtDepth[...](0xc027fd0a40, {0xc002a81d00, 0x3}, 0x1)
Apr 18 15:58:33 stks03 k3s[1785244]:         /go/pkg/mod/github.com/libp2p/go-libp2p-kad-dht@v0.36.0/provider/internal/keyspace/trie.go:259 +0xf6
Apr 18 15:58:33 stks03 k3s[1785244]: github.com/libp2p/go-libp2p-kad-dht/provider/internal/keyspace.pruneSubtrieAtDepth[...](0xc0006165c0, {0xc002a81d00, 0x3}, 0x0)
Apr 18 15:58:33 stks03 k3s[1785244]:         /go/pkg/mod/github.com/libp2p/go-libp2p-kad-dht@v0.36.0/provider/internal/keyspace/trie.go:259 +0xf6
Apr 18 15:58:33 stks03 k3s[1785244]: github.com/libp2p/go-libp2p-kad-dht/provider/internal/keyspace.PruneSubtrie[...](...)
Apr 18 15:58:33 stks03 k3s[1785244]:         /go/pkg/mod/github.com/libp2p/go-libp2p-kad-dht@v0.36.0/provider/internal/keyspace/trie.go:241
Apr 18 15:58:33 stks03 k3s[1785244]: github.com/libp2p/go-libp2p-kad-dht/provider/internal/timeseries.(*CycleStats).Add(0xc001e3a790, {0xc002a81d00, 0x3}, 0x45581b)
Apr 18 15:58:33 stks03 k3s[1785244]:         /go/pkg/mod/github.com/libp2p/go-libp2p-kad-dht@v0.36.0/provider/internal/timeseries/cycle_stats.go:61 +0xa6
Apr 18 15:58:33 stks03 k3s[1785244]: github.com/libp2p/go-libp2p-kad-dht/provider.(*SweepingProvider).batchReprovide.func1()
Apr 18 15:58:33 stks03 k3s[1785244]:         /go/pkg/mod/github.com/libp2p/go-libp2p-kad-dht@v0.36.0/provider/provider.go:1734 +0x76
Apr 18 15:58:33 stks03 k3s[1785244]: github.com/libp2p/go-libp2p-kad-dht/provider.(*SweepingProvider).batchReprovide(0xc001e3a488, {0xc002a81d00, 0x3})
Apr 18 15:58:33 stks03 k3s[1785244]:         /go/pkg/mod/github.com/libp2p/go-libp2p-kad-dht@v0.36.0/provider/provider.go:1748 +0x37e
Apr 18 15:58:33 stks03 k3s[1785244]: github.com/libp2p/go-libp2p-kad-dht/provider.(*SweepingProvider).reprovideLateRegions.func1({0xc002a81d00?, 0x22?})
Apr 18 15:58:33 stks03 k3s[1785244]:         /go/pkg/mod/github.com/libp2p/go-libp2p-kad-dht@v0.36.0/provider/provider.go:1646 +0x2a
Apr 18 15:58:33 stks03 k3s[1785244]: created by github.com/libp2p/go-libp2p-kad-dht/provider.(*SweepingProvider).reprovideLateRegions in goroutine 5877764
Apr 18 15:58:33 stks03 k3s[1785244]:         /go/pkg/mod/github.com/libp2p/go-libp2p-kad-dht@v0.36.0/provider/provider.go:1645 +0x210
Apr 18 15:58:34 stks03 systemd[1]: k3s.service: Main process exited, code=exited, status=2/INVALIDARGUMENT
```

We have confirmed that the panic is often preceded by Raft/etcd latency spikes (>100ms). This suggests that as the Go scheduler is pressured, the timing gap between trie read/write operations increases, making the race condition much more likely to trigger.

--- a/provider/provider.go
+++ b/provider/provider.go
@@ -1795,7 +1795,9 @@ func (s *SweepingProvider) batchReprovide(prefix bitstr.Key) {
 	defer func() {
 		elapsed := time.Since(startTime)
 		s.stats.ongoingReprovides.finish(len(keys))
+		s.stats.cycleStatsLk.Lock()
 		s.stats.reprovideDuration.Add(prefix, int64(elapsed))
+		s.stats.cycleStatsLk.Unlock()
 		s.logger.Infof("reprovide finished for prefix %q, %d keys in %s", prefix, keyCount, elapsed)
 	}()
 

--- a/provider/stats_test.go
+++ b/provider/stats_test.go
@@ -9,6 +9,9 @@ import (
 	"testing/synctest"
 	"time"
 
+	"github.com/ipfs/go-datastore"
+	"github.com/ipfs/go-datastore/namespace"
+	dssync "github.com/ipfs/go-datastore/sync"
 	"github.com/ipfs/go-libdht/kad/key"
 	"github.com/ipfs/go-libdht/kad/key/bit256"
 	"github.com/ipfs/go-libdht/kad/key/bitstr"
@@ -749,6 +752,108 @@ func TestStatsContextDeadline(t *testing.T) {
 		require.Emptyf(t, s, "Stats should return a zero snapshot when ctx is done")
 		require.GreaterOrEqualf(t, elapsed, deadline, "Stats should not return before the deadline elapses")
 		require.Lessf(t, elapsed, deadline+time.Second, "Stats should return shortly after the deadline elapses")
+	})
+}
+
+// TestStatsConcurrentReprovideRace is a regression test for
+// libp2p/go-libp2p-kad-dht#1254 (fixed in #1255). It drives many concurrent
+// batchReprovide invocations alongside repeated Stats() reads to lock in the
+// invariant that every CycleStats access (reprovideDuration, keysPerReprovide,
+// peers, reachable) runs under cycleStatsLk. Run with -race.
+func TestStatsConcurrentReprovideRace(t *testing.T) {
+	synctest.Test(t, func(t *testing.T) {
+		ctx := t.Context()
+
+		pid, err := peer.Decode("12D3KooWCPQTeFYCDkru8nza3Af6u77aoVLA71Vb74eHxeR91Gka")
+		require.NoError(t, err)
+
+		replicationFactor := 2
+		peerPrefixBitlen := 4
+		require.LessOrEqual(t, peerPrefixBitlen, bitsPerByte)
+		var nPeers byte = 1 << peerPrefixBitlen
+		peers := make([]peer.ID, nPeers)
+		for i := range nPeers {
+			b := i << (bitsPerByte - peerPrefixBitlen)
+			k := [32]byte{b}
+			peers[i], err = kb.GenRandPeerIDWithCPL(k[:], uint(peerPrefixBitlen))
+			require.NoError(t, err)
+		}
+
+		router := &mockRouter{
+			getClosestPeersFunc: func(_ context.Context, k string) ([]peer.ID, error) {
+				sortedPeers := kb.SortClosestPeers(peers, kb.ConvertKey(k))
+				return sortedPeers[:min(replicationFactor, len(peers))], nil
+			},
+		}
+		msgSender := &mockMsgSender{
+			sendMessageFunc: func(context.Context, peer.ID, *pb.Message) error { return nil },
+		}
+
+		ds := dssync.MutexWrap(datastore.NewMapDatastore())
+		ks, err := keystore.NewKeystore(ds)
+		require.NoError(t, err)
+		defer ks.Close()
+
+		// 32 multihashes spanning 32 distinct 5-bit prefixes. Each
+		// batchReprovide(prefix) call retrieves a single key, falls
+		// through to individualProvide -> vanillaProvide (acquiring
+		// cycleStatsLk at provider.go:835), and then runs the defer
+		// that writes to reprovideDuration (the lock at provider.go:1799
+		// that this regression test exists to protect).
+		const prefixLen = 5
+		mhs := genBalancedMultihashes(prefixLen)
+		_, err = ks.Put(ctx, mhs...)
+		require.NoError(t, err)
+
+		prefixes := make([]bitstr.Key, 0, len(mhs))
+		for _, h := range mhs {
+			prefixes = append(prefixes, bitstr.Key(key.BitString(keyspace.MhToBit256(h))[:prefixLen]))
+		}
+
+		provDs := namespace.Wrap(ds, datastore.NewKey("provider"))
+		prov, err := New(
+			WithReprovideInterval(time.Hour),
+			WithReplicationFactor(replicationFactor),
+			WithPeerID(pid),
+			WithRouter(router),
+			WithMessageSender(msgSender),
+			WithSelfAddrs(getMockAddrs(t)),
+			WithKeystore(ks),
+			WithDatastore(provDs),
+			WithSkipBootstrapReprovide(true),
+		)
+		require.NoError(t, err)
+		defer prov.Close()
+
+		synctest.Wait()
+		require.True(t, prov.connectivity.IsOnline())
+
+		// Reader: hammer Stats() while reprovides run.
+		readerCtx, cancelReader := context.WithCancel(ctx)
+		readerDone := make(chan struct{})
+		go func() {
+			defer close(readerDone)
+			for readerCtx.Err() == nil {
+				_, _ = prov.Stats(readerCtx)
+			}
+		}()
+
+		// Writers: many concurrent batchReprovides across distinct prefixes,
+		// repeated a few times to widen the race window.
+		var wg sync.WaitGroup
+		for range 4 {
+			for _, prefix := range prefixes {
+				wg.Add(1)
+				go func(p bitstr.Key) {
+					defer wg.Done()
+					prov.batchReprovide(p)
+				}(prefix)
+			}
+		}
+		wg.Wait()
+
+		cancelReader()
+		<-readerDone
 	})
 }
 

